modular/executor/execute_gc.go

```diff
@@ -0,0 +1,545 @@
+		log.CtxErrorw(ctx, "failed to query gc integrity pieces list", "task_info", task.Info(), "error", err)
+		return err
+	}
+
+	if len(waitingVerifyGCIntegrityPieces) == 0 {
+		log.CtxInfow(ctx, "no waiting gc integrity pieces", "task_info", task.Info())
+		return nil
+	}
+
+	for _, integrityObject := range waitingVerifyGCIntegrityPieces {
+		log.CtxDebugw(ctx, "gc zombie current waiting verify gc integrity meta piece", "integrity_meta", integrityObject)
+		objID := integrityObject.ObjectID
+		// If metadata service has object info, use objInfoFromMetaData, otherwise query from chain
+		if objInfoFromMetaData, err = e.baseApp.GfSpClient().GetObjectByID(ctx, objID); err != nil {
+			log.Errorf("failed to get object meta from meta data service, will check from chain", "error", err)
+			if strings.Contains(err.Error(), "no such object from metadata") {
+				// If deletion is possible, has integrity hash, lacks object info, perform a chain check, and verify against the chain.
+				if objInfoFromChain, err = e.baseApp.Consensus().QueryObjectInfoByID(ctx, strconv.FormatUint(objID, 10)); err != nil {
+					if strings.Contains(err.Error(), "No such object") {
+						// 1) This object does not exist on the chain
+						e.gcWorker.deleteObjectPiecesAndIntegrityMeta(ctx, integrityObject)
+					}
+				} else {
+					// 2) query metadata error, but chain has the object info, gvg  primary sp should have integrity meta
+					if e.gcWorker.checkGVGMatchSP(ctx, objInfoFromChain, integrityObject.RedundancyIndex) == ErrInvalidRedundancyIndex {
+						e.gcWorker.deleteObjectPiecesAndIntegrityMeta(ctx, integrityObject)
+					}
+					continue
+				}
+			}
+		} else {
+			// 3) check integrity meta & object info
+			if e.gcWorker.checkGVGMatchSP(ctx, objInfoFromMetaData, integrityObject.RedundancyIndex) == ErrInvalidRedundancyIndex {
+				e.gcWorker.deleteObjectPiecesAndIntegrityMeta(ctx, integrityObject)
+			}
+		}
+	}
+	return nil
+}
+
+func (e *ExecuteModular) gcZombiePieceFromPieceHash(ctx context.Context, task coretask.GCZombiePieceTask) error {
+	var (
+		err                    error
+		waitingVerifyGCPieces  []*spdb.GCPieceMeta
+		objectInfoFromMetadata *storagetypes.ObjectInfo
+		objInfoFromChain       *storagetypes.ObjectInfo
+	)
+	log.CtxDebugw(ctx, "start to gc zombie piece from piece hash", "task_info", task.Info())
+
+	// replicate piece checksum must on secondary sp
+	if waitingVerifyGCPieces, err = e.baseApp.GfSpDB().ListReplicatePieceChecksumByObjectIDRange(
+		int64(task.GetStartObjectId()), int64(task.GetEndObjectId())); err != nil {
+		log.CtxErrorw(ctx, "failed to query replicate piece checksum", "task_info", task.Info(), "error", err)
+		return err
+	}
+
+	if len(waitingVerifyGCPieces) == 0 {
+		log.CtxErrorw(ctx, "no waiting gc pieces", "task_info", task.Info())
+		return nil
+	}
+
+	for _, piece := range waitingVerifyGCPieces {
```

Comment:
a object might have multiple piece hash records, seems the call to chain or meta are duplicate.